src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -295,6 +325,98 @@ void ScheduleTask(DateTimeOffset utcNow)
             }
         }
 
+        private readonly List<WeakReference<Stats>?>? _allStats;
+        private readonly Stats? _accumulatedStats;
+        private ThreadLocal<Stats>? _stats;
+
+        private (long, long) Sum()
+        {
+            long hits = 0, misses = 0;
+            lock (_allStats!)
+            {
+                hits += _accumulatedStats!.Hits;
+                misses += _accumulatedStats!.Misses;
+                foreach (WeakReference<Stats>? wr in _allStats!)
+                {
+                    if (wr is not null && wr.TryGetTarget(out Stats? stats))
+                    {
+                        hits += Interlocked.Read(ref stats.Hits);
+                        misses += Interlocked.Read(ref stats.Misses);
+                    }
+                }
+
+                _allStats.TrimExcess();
+            }
+            return (hits, misses);
+        }
+
+        private void Hit()
+        {
+            if (_allStats is not null)
+            {
+                if (IntPtr.Size == 4)
+                    Interlocked.Increment(ref GetStats().Hits);
+                else
+                    GetStats().Hits++;
+            }
+        }
+
+        private void Miss()
+        {
+            if (_allStats is not null)
+            {
+                if (IntPtr.Size == 4)
+                    Interlocked.Increment(ref GetStats().Misses);
+                else
+                    GetStats().Misses++;
+            }
+        }
+
+        private Stats GetStats() => _stats!.Value!;
+
+        internal sealed class Stats
+        {
+            public long Hits;
+            public long Misses;
+            private readonly MemoryCache _memoryCache;
+            public Stats(MemoryCache memoryCache, bool skipAdd = false)
+            {
+                _memoryCache = memoryCache;
+                if (!skipAdd)
+                    _memoryCache.AddToStats(this);
+            }
+            ~Stats()
+            {
+                _memoryCache.RemoveFromStats(this);
```

Comment:
If we never added this instance to the _memoryCache, should we also not remove from it?

Similarly, I wonder if the reference to the `MemoryCache` shouldn't be a WeakReference. Is there a case where one of these ThreadLocals could live longer than the MemoryCache, and it would be a memory leak?